### PR TITLE
Change 'pl-number-input' to accept fractional answers

### PR DESCRIPTION
**Pull request by henryg3**
Added the ability to use fractional inputs by checking if slash is present in string. If present, the quotient will be computed and then sent as an answer. Adding fractional input functionality will be useful in the event a student does not have a calculator to convert a difficult fraction to decimal form during a test or if they're studying.

### CODE REVIEW

File: elements/pl-number-input/pl-number-input.py [hunk 1]
```diff
@@ -183,21 +183,51 @@ def parse(element_html, data):
...
-        a_sub_parsed = pl.string_to_number(a_sub, allow_complex=allow_complex)
-        if a_sub_parsed is None:
-            raise ValueError('invalid submitted answer (wrong type)')
-        if not np.isfinite(a_sub_parsed):
-            raise ValueError('invalid submitted answer (not finite)')
+        # Accept fractional input, then convert it to a float or decimal
+        # (only works for certain integer addition/subtraction problems)
+    elif '/' in a_sub:
+        a_sub = a_sub.strip()
+        a_sub_list = a_sub.split('/')
```
Comment (mwest1066) [inline elements/pl-number-input/pl-number-input.py]: What if there is more than one `'/'` character in the string?

File: elements/pl-number-input/pl-number-input.py [hunk 2]
```diff
@@ -183,21 +183,51 @@ def parse(element_html, data):
...
-    try:
-        a_sub_parsed = pl.string_to_number(a_sub, allow_complex=allow_complex)
-        if a_sub_parsed is None:
-            raise ValueError('invalid submitted answer (wrong type)')
-        if not np.isfinite(a_sub_parsed):
-            raise ValueError('invalid submitted answer (not finite)')
+        # Accept fractional input, then convert it to a float or decimal
+        # (only works for certain integer addition/subtraction problems)
+    elif '/' in a_sub:
+        a_sub = a_sub.strip()
```
Comment (mwest1066) [inline elements/pl-number-input/pl-number-input.py]: is this `strip()` necessary?

File: elements/pl-number-input/pl-number-input.py [hunk 3]
```diff
@@ -183,21 +183,51 @@ def parse(element_html, data):
...
+        a_sub_list = a_sub.split('/')
+        try:
+            a_sub_parsed_numerator = pl.string_to_number(a_sub_list[0])
+            if a_sub_parsed_numerator is None:
+                raise ValueError('invalid submitted answer (wrong type)')
+            if not np.isfinite(a_sub_parsed_numerator):
+                raise ValueError('invalid submitted answer (not finite)')
+                return
```
Comment (mwest1066) [inline elements/pl-number-input/pl-number-input.py]: why do we have a `return` after a `raise`?

File: elements/pl-number-input/pl-number-input.py [hunk 4]
```diff
@@ -183,21 +183,51 @@ def parse(element_html, data):
...
+        except Exception:
+            data['format_errors'][name] = 'Invalid format. The submitted answer could not be interpreted as a double-precision floating-point number or fraction.'
+            data['submitted_answers'][name] = None
+            return
+        try:
+            a_sub_parsed_denominator = pl.string_to_number(a_sub_list[1])
+            if a_sub_parsed_denominator is None:
+                raise ValueError('invalid submitted answer (wrong type)')
+            if not np.isfinite(a_sub_parsed_denominator):
+                raise ValueError('invalid submitted answer (not finite)')
```
Comment (mwest1066) [inline elements/pl-number-input/pl-number-input.py]: I'm having trouble following the Exception logic here -- it looks like we raise a ValueError with an error message, but then we catch it below and discard this error message?

File: elements/pl-number-input/pl-number-input.mustache [hunk 5]
```diff
@@ -71,7 +71,7 @@ Your answer must be either a real number between <code>-1e308</code> and <code>1
...
+Fractions (e.g., <code>3/4</code> and <code>1.0/2.5</code>) are accepted, while no other symbolic expressions are accepted. Scientific notation is accepted (e.g., <code>1.2e03</code>). {{#relabs}}<p>Your answer must be accurate to within relative tolerance <code>{{rtol}}</code> and absolute tolerance <code>{{atol}}</code>.</p>{{/relabs}}{{#sigfig}}<p>Your answer must be accurate to {{digits}} significant figures. This means (for example) that if the true answer is <code>1.234</code>, then the submitted answer must be between <code>1.234 - {{comparison_eps}}</code> and <code>1.234 + {{comparison_eps}}</code> to be counted correct.</p>{{/sigfig}}{{#decdig}}<p>Your answer must be accurate to {{digits}} digits after the decimal. This means (for example) that if the true answer is <code>1.234</code>, then the submitted answer must be between <code>1.234 - {{comparison_eps}}</code> and <code>1.234 + {{comparison_eps}}</code> to be counted correct.</p>{{/decdig}}
```
Comment (mwest1066) [inline elements/pl-number-input/pl-number-input.mustache]: How do fractions interact with complex numbers? Do we handle fractions in complex numbers? If not, we should tell the user this in the help text.

File: elements/pl-number-input/pl-number-input.py [hunk 6]
```diff
@@ -183,21 +183,51 @@ def parse(element_html, data):
...
-            raise ValueError('invalid submitted answer (not finite)')
+        # Accept fractional input, then convert it to a float or decimal
+        # (only works for certain integer addition/subtraction problems)
+    elif '/' in a_sub:
+        a_sub = a_sub.strip()
+        a_sub_list = a_sub.split('/')
+        try:
+            a_sub_parsed_numerator = pl.string_to_number(a_sub_list[0])
+            if a_sub_parsed_numerator is None:
+                raise ValueError('invalid submitted answer (wrong type)')
```
Comment (mwest1066) [inline elements/pl-number-input/pl-number-input.py]: In these errors, should we specify that it's the numerator that's wrong? And similarly for other errors.

File: elements/pl-number-input/pl-number-input.py (hunk 1 above)
Comment (henryg3) [reply elements/pl-number-input/pl-number-input.py]: There will be an error message that shows after the user grades their answer 
screen shot 2019-02-23 at 5 44 02 pm

File: elements/pl-number-input/pl-number-input.py (hunk 2 above)
Comment (henryg3) [reply elements/pl-number-input/pl-number-input.py]: I just realized .split() makes a separated list, so it is not necessary. Removed .strip() in the code.

File: elements/pl-number-input/pl-number-input.mustache (hunk 5 above)
Comment (henryg3) [reply elements/pl-number-input/pl-number-input.mustache]: They should not be handled with complex numbers. Noted.

File: elements/pl-number-input/pl-number-input.py (hunk 3 above)
Comment (henryg3) [reply elements/pl-number-input/pl-number-input.py]: Just noticed the redundancy. Noted and removed.

File: elements/pl-number-input/pl-number-input.py (hunk 6 above)
Comment (henryg3) [reply elements/pl-number-input/pl-number-input.py]: Noted and added.

Comment (henryg3) [reply elements/pl-number-input/pl-number-input.py]: For the denominator, what would we do in the case that it is a large, near-infinite value?

Comment (henryg3) [reply elements/pl-number-input/pl-number-input.py]: If the denominator was an infinitely large value, wouldn't the fraction simply equate to 0?

**Comment by mwest1066** [comment]
I'm testing this on the `addNumbers` question in the example course. Entering `1/0` causes an uncaught Exception to be thrown.

File: elements/pl-number-input/pl-number-input.py (hunk 1 above)
Comment (mwest1066) [reply elements/pl-number-input/pl-number-input.py]: Entering `20/5/10` is accepted as a valid number, which it should not be.

**Comment by mwest1066** [comment]
Entering `1e200/1e-200` throws an uncaught exception.

**Comment by mwest1066** [comment]
@henryg3 did you want to continue working on this PR? It's fine if you don't have time, and in that case I'll go ahead and close it out.

**Comment by henryg3** [comment]
> @henryg3 did you want to continue working on this PR? It's fine if you don't have time, and in that case I'll go ahead and close it out.

Hi Matt, I've been really busy with schoolwork and midterms next week, so I have not been able to continue working on this PR, but I am still interested in working on it, perhaps over the summer.

**Comment by mwest1066** [comment]
Thanks Henry! I'll close this PR for now, but we can always re-open it again if you have time to work on it. Just let me know!

**Pull request closed without merging**